### PR TITLE
Implement PlainTime

### LINKED ISSUE #29: balanceDuration() is broken when when largestUnit is greater than TimeComponent.days
https://github.com/ColinEberhardt/assemblyscript-temporal/blob/1d02f91541d1090e8bbf9650b6f5a793f91a84b7/assembly/utils.ts#L389-L398

When `largestUnit` is is greater than `TimeComponent.days` then,
`nanoseconds` variable is not updated (ideally should be updated to `durationNs`)

So for `TimeComponent.hours`, `TimeComponent.minutes `,etc it will always return empty Duration `PT0S` if nanosecond is 0 else just the nanosecond.

See the failing tests in https://github.com/ColinEberhardt/assemblyscript-temporal/runs/2269898985

cc @MaxGraey

### PR DESCRIPTION
**Pull request by ashutoshvarma**
Part of #21

- [ ] PlainTime
  - [x] compare (https://github.com/ColinEberhardt/assemblyscript-temporal/pull/8)
  - [x] equals (https://github.com/ColinEberhardt/assemblyscript-temporal/pull/8)
  - [x] from
  - [x] add
  - [x] subtract
  - [x] toString
  - [x] with
  - [x] until
  - [x] since
  - [ ] round
  - [ ] toJSON
  - [ ] valueOf
  - [x] toPlainDateTime

Other fixes include in PR
- fix PlainDateTime.fromString for "1976-11-18T15:23:30.1Z" family              [[2d0abd5](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/2d0abd54b9dd64c74271f1393d87db3d4286c59a)]
- fix PlainDateTime.fromString for "1976-11-18T152330.1+00:00" family      [[73b57ec](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/73b57ec507ace84a40fc05fc6e57d70ce2a1af78)]
- fix #29 : balanceDuration() for largestUnit greater than Days                 [[5c36d22](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/5c36d2263079106350570bb89c3c9f5d07756c71)]
- fix Duration.toString() for negative values [[6cc1621](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/6cc16218dee06247993293e70fe657478de05468)]
- fix trailing zero in Duration.toString()

### CODE REVIEW

**Comment by ashutoshvarma** [comment]
From failing tests log, seems FP rounding error
```log
[Fail]: ✖ 15:23:30.123456789.add({ nanoseconds: 300 })
[Actual]: "15:23:30.12345708899999999"
[Expected]: "15:23:30.123457089"
```

How should we handle floating-point rounding errors?

File: assembly/plaintime.ts [hunk 1]
```diff
@@ -14,33 +32,161 @@ export class PlainTime {
...
+
+  toString(): string {
+    // 22:54:31
+    return (
+      toPaddedString(this.hour) +
+      ":" +
+      toPaddedString(this.minute) +
+      ":" +
+      toPaddedString(this.second) +
+      (this.nanosecond != 0 || this.microsecond != 0 || this.millisecond != 0
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
      ((this.nanosecond | this.microsecond | this.millisecond) != 0
```

File: assembly/plaintime.ts [hunk 2]
```diff
@@ -64,4 +210,86 @@ export class PlainTime {
 
     return ord(a.nanosecond, b.nanosecond);
   }
+
+  add<T>(durationToAdd: T): PlainTime {
+    const duration =
+      durationToAdd instanceof DurationLike
+        ? durationToAdd.toDuration()
+        : // @ts-ignore TS2352
+          (durationToAdd as Duration);
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
        // @ts-ignore TS2352
        : (durationToAdd as Duration);
```

File: assembly/plaintime.ts [hunk 3]
```diff
@@ -64,4 +210,86 @@ export class PlainTime {
...
+      newTime.nanosecond
+    );
+  }
+
+  subtract<T>(durationToSubtract: T): PlainTime {
+    const duration =
+      durationToSubtract instanceof DurationLike
+        ? durationToSubtract.toDuration()
+        : // @ts-ignore TS2352
+          (durationToSubtract as Duration);
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
        // @ts-ignore TS2352
        : (durationToSubtract as Duration);
```

**Comment by MaxGraey** [comment]
> How should we handle floating-point rounding errors?

Such imprecise rarely happens in `dtoa` implementaion in AS. I have plan rewrite that to more modern Ryu's implementation. But currently we could just ignore this imprecision and add some TODO comment

File: assembly/plaintime.ts [hunk 4]
```diff
@@ -64,4 +238,86 @@ export class PlainTime {
 
     return ord(a.nanosecond, b.nanosecond);
   }
+
+  add<T>(durationToAdd: T): PlainTime {
+    const duration =
+      durationToAdd instanceof DurationLike
+        ? durationToAdd.toDuration()
+        : // @ts-ignore TS2352
+          (durationToAdd as Duration);
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
        // @ts-ignore TS2352
        : (durationToAdd as Duration);
```

File: assembly/plaintime.ts [hunk 5]
```diff
@@ -64,4 +238,86 @@ export class PlainTime {
...
+      newTime.nanosecond
+    );
+  }
+
+  subtract<T>(durationToSubtract: T): PlainTime {
+    const duration =
+      durationToSubtract instanceof DurationLike
+        ? durationToSubtract.toDuration()
+        : // @ts-ignore TS2352
+          (durationToSubtract as Duration);
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
        // @ts-ignore TS2352
        : (durationToSubtract as Duration);
```

File: assembly/plaintime.ts (hunk 4 above)
Comment (ashutoshvarma) [reply assembly/plaintime.ts]: Ohh crap, I should disable eslint in vscode 😅

File: assembly/plaintime.ts [hunk 6]
```diff
@@ -14,33 +34,187 @@ export class PlainTime {
...
+            f64(this.millisecond) / 1_000.0
+          )
+            .toString()
+            .substring(1)
+        : "")
+    );
+  }
+
+  toPlainDateTime(
+    dateLike: DateLike = { year: 0, month: 0, day: 0 }
```
Comment (MaxGraey) [inline assembly/plaintime.ts]: According to this:
https://github.com/tc39/proposal-temporal/blob/2f36d91332aeebc3a66082600c8eef57b751fbd5/polyfill/test/PlainDate/prototype/toPlainDateTime/time-undefined.js#L10

better use this approach:
```ts
toPlainDateTime(dateLike: DateLike | null = null): PlainDateTime {
  let year = 0, month = 0, day = 0;
  if (dateLike !== null) {
     year  = coalesce(dateLike!.year, 0);
     month = coalesce(dateLike!.month, 0);
     day   = coalesce(dateLike!.day, 0);
  }
  return new PlainDateTime(
    year,
    month,
    day,
    this.hour,
    this.minute,
    this.second,
    this.millisecond,
    this.microsecond,
    this.nanosecond
  );
}
```

Comment (ashutoshvarma) [reply assembly/plaintime.ts]: Done 👍

File: assembly/plaintime.ts [hunk 7]
```diff
@@ -64,4 +242,86 @@ export class PlainTime {
 
     return ord(a.nanosecond, b.nanosecond);
   }
+
+  add<T>(durationToAdd: T): PlainTime {
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
  add<T = DurationLike>(durationToAdd: T): PlainTime {
```

File: assembly/plaintime.ts [hunk 8]
```diff
@@ -64,4 +242,86 @@ export class PlainTime {
...
+      newTime.hour,
+      newTime.minute,
+      newTime.second,
+      newTime.millisecond,
+      newTime.microsecond,
+      newTime.nanosecond
+    );
+  }
+
+  subtract<T>(durationToSubtract: T): PlainTime {
```
Comment (MaxGraey) [inline assembly/plaintime.ts]:
```suggestion
  subtract<T = DurationLike>(durationToSubtract: T): PlainTime {
```

File: assembly/plaintime.ts (hunk 7 above)
Comment (MaxGraey) [reply assembly/plaintime.ts]: And after this you could remove all type hints in tests like 
`.add<DurationLike>({ hours: 2, ... })` to `.add({ hours: 2, ... })`

Comment (ashutoshvarma) [reply assembly/plaintime.ts]: Done

File: assembly/plaintime.ts (hunk 8 above)
Comment (ashutoshvarma) [reply assembly/plaintime.ts]: Done

**Comment by ashutoshvarma** [comment]
@MaxGraey shouldn't it be 8 instead of 4 in `sign` ? or am I missing something?
https://github.com/ColinEberhardt/assemblyscript-temporal/blob/1d02f91541d1090e8bbf9650b6f5a793f91a84b7/assembly/utils.ts#L177-L181

**Comment by MaxGraey** [comment]
right! Good catch!

**Comment by ashutoshvarma** [comment]
Also while working on `util()`, I get to know that `balanceDuration` is broken when `largestUnit` is greater than `TimeComponent.days` .

I will create an issue for it shortly

**Comment by ashutoshvarma** [comment]
> right! Good catch!

Should I add a commit for it in this PR ? `PlainTime.until()` tests are failing because of it too

**Comment by MaxGraey** [comment]
I already made PR with fix: https://github.com/ColinEberhardt/assemblyscript-temporal/pull/28

File: assembly/utils.ts [hunk 9]
```diff
@@ -395,83 +395,87 @@ export function balanceDuration(
...
   }
 
-  const sig = sign(nanoseconds);
-  nanoseconds = abs(nanoseconds);
-  microseconds = 0;
-  milliseconds = 0;
-  seconds = 0;
-  minutes = 0;
-  hours = 0;
+  const sig = sign(nanosecondsI64);
```
Comment (MaxGraey) [inline assembly/utils.ts]:
```suggestion
  const sig = i32(sign(nanosecondsI64));
```

File: assembly/utils.ts [hunk 10]
```diff
@@ -482,13 +486,13 @@ export function balanceDuration(
...
-    milliseconds * sig,
-    microseconds * sig,
-    nanoseconds * sig
+    i32(daysI64),
+    i32(hoursI64 * sig),
+    i32(minutesI64 * sig),
+    i32(secondsI64 * sig),
+    i32(millisecondsI64 * sig),
+    i32(microsecondsI64 * sig),
+    i32(nanosecondsI64 * sig)
```
Comment (MaxGraey) [inline assembly/utils.ts]:
```suggestion
    i32(daysI64),
    i32(hoursI64) * sig,
    i32(minutesI64) * sig,
    i32(secondsI64) * sig,
    i32(millisecondsI64) * sig,
    i32(microsecondsI64) * sig,
    i32(nanosecondsI64) * sig
```

File: assembly/utils.ts (hunk 9 above)
Comment (ashutoshvarma) [reply assembly/utils.ts]: Does casting sig to `i32` change something?

Comment (MaxGraey) [reply assembly/utils.ts]: See next suggestion. Main idea is avoiding 64-bit arithmetic as much as possible due to some wasm engines or some 32-bit platforms may be significantly slower with 64-bit arithmetic

Comment (ashutoshvarma) [reply assembly/utils.ts]: Ohh, I see.

**Comment by ashutoshvarma** [comment]
https://github.com/ColinEberhardt/assemblyscript-temporal/blob/1d02f91541d1090e8bbf9650b6f5a793f91a84b7/assembly/duration.ts#L71-L78

@MaxGraey is there a way to remove the trailing `.0` from the f64 string representation? I tried reparsing string back using `parseFloat` and then `toString` but it didn't work

From CI logs https://github.com/ColinEberhardt/assemblyscript-temporal/runs/2270205795
```
[Fail]: ✖ the default largest unit is at least hours
[Actual]: "PT6H52M42.0S"
[Expected]: "PT6H52M42S"
```

**Comment by MaxGraey** [comment]
> is there a way to remove the trailing .0 from the f64 string representation? I tried reparsing string back using 

Trailing zero for serialized f64/f32 strings is a deliberate decision and convention of AssemblyScript which differ from JS/TS. So I suggest slightly modify original tests and add `.0` when this required

**Comment by ashutoshvarma** [comment]
I think this PR is good to go now!

**Comment by MaxGraey** [comment]
> > is there a way to remove the trailing .0 from the f64 string representation? I tried reparsing string back using
> 
> Trailing zero for serialized f64/f32 strings is a deliberate decision and convention of AssemblyScript which differ from JS/TS. So I suggest slightly modify original tests and add `.0` when this required

Another option:
```ts
function stringify(value: f64): string {
   return F64.isSafeInteger(value) ? i64(value).toString() : value.toString();
}
```

**Comment by ashutoshvarma** [comment]
> ```ts
> function stringify(value: f64): string {
>    return F64.isSafeInteger(value) ? i64(value).toString() : value.toString();
> }
> ```

This is great! With this `Duration.toString(`) will be more close to tc39 implementation

**Comment by ColinEberhardt** [comment]
> I think this PR is good to go now!

Thank you - great work 👍

**Pull request merged**